datalad/utils.py

```diff
@@ -2590,3 +2591,41 @@ def check_symlink_capability(path, target):
             path.unlink()
         if target.exists():
             target.unlink()
+
+
+_cp_supports_reflink = None
+
+
+def copy_file(src, dst, follow_symlinks=True):
+    """
+    Copy file (including metadata) from src to dst.  Use `cp --reflink=auto` if
+    the local `cp` supports it.
+
+    Parameters
+    ----------
+    src: str or os.PathLike
+      path to the file to copy
+    dst: str or os.PathLike
+      path that `src` should be copied to
+    """
+    global _cp_supports_reflink
+    if _cp_supports_reflink is None:
+        r = subprocess.run(
+            ["cp", "--help"],
+            stdout=subprocess.PIPE,
+            stderr=subprocess.STDOUT,
+            universal_newlines=True,
+        )
+        # Ignore command failures (e.g., if cp doesn't support --help), as the
+        # command will still likely output its usage info.
+        _cp_supports_reflink = "--reflink" in r.stdout
+    if _cp_supports_reflink:
+        symopts = ["-R", "-L"] if follow_symlinks else ["-R", "-P"]
+        subprocess.run(
+            ["cp", "-f", "-p", "--reflink=auto"]
+            + symopts
+            + ["--", str(src), str(dst)],
+            check=True,
+        )
+    else:
+        return shutil.copy2(src, dst, follow_symlinks=follow_symlinks)
```

Comment:
haven't looked inside shutil.copy2 etc, is there any way/hope to get some alignment in exceptions which would be raised if e.g. permissions denied or dst directory does not exist?  
I do not think we should "parse" the output from `cp` call but ideally there should be at least some common base exception or smth like that and if different (which is likely) between two possible runs -- docstring adjusted to document that.

Reply:
No.  If `cp` fails, all the caller gets is a `CalledProcessError` with the exit code (which isn't standardized) and the output.

Reply:
and indeed there are dedicated exceptions by copy2:
```
> python -c 'import shutil; shutil.copyfile(".git", "/tmp/gggg")'
Traceback (most recent call last):
  File "<string>", line 1, in <module>
  File "/usr/lib/python3.9/shutil.py", line 264, in copyfile
    with open(src, 'rb') as fsrc, open(dst, 'wb') as fdst:
IsADirectoryError: [Errno 21] Is a directory: '.git'

$> python -c 'import shutil; shutil.copy2(".git/index", "/")' 
Traceback (most recent call last):
  File "<string>", line 1, in <module>
  File "/usr/lib/python3.9/shutil.py", line 435, in copy2
    copyfile(src, dst, follow_symlinks=follow_symlinks)
  File "/usr/lib/python3.9/shutil.py", line 264, in copyfile
    with open(src, 'rb') as fsrc, open(dst, 'wb') as fdst:
PermissionError: [Errno 13] Permission denied: '/index'
```
I am ok to sacrifice them .... but maybe others not? @datalad/developers ?

Reply:
We are not monkey-patching things, so I am OK with that. But maybe add a `Raises` section to the docstring that states what would happen. DataLad consumers might expect a `CommandError` (which is not coming), and Python consumers might expect the normal type of exceptions (which will not happen either).

Reply:
Docstring updated.